pyspectrafuse/commands/build_cluster_db.py

```
@@ -0,0 +1,450 @@
+    try:
+        conn.execute("SET memory_limit='4GB'")
+        conn.execute(f"SET temp_directory='{tmp_dir}'")
+        conn.execute("SET threads=4")
+    except Exception:
+        conn.close()
+        raise
+
+    # Build the source parquet UNION with column normalization
+    source_parts = []
+    for ppath in all_parquet_files:
+        cols_df = conn.execute(f"SELECT name FROM parquet_schema('{ppath}')").fetchdf()
```

Ruff (0.15.10)

[error] Possible SQL injection vector through string-based query construction

(S608)